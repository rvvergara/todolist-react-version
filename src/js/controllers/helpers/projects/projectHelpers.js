import uuid from 'uuid';

import projectsController from '../../projectsController';

import * as localStorageData from '../common/storage';

export const createProjectsArray = () => {
  if (localStorageData.getDataFromLocalStorage('projectsArray') === null) {
    const projectsArray = [];
    localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
  }
};

// Create a default project for new users or after localstorage is cleared
export const createDefaultProject = () => {
  if (localStorageData.getDataFromLocalStorage('projectCount') === 0) {
    const defaultProject = projectsController.create('Default Project', uuid());
  }
  if (localStorageData.getDataFromLocalStorage('Default Project') !== null) {
    const defaultProject = localStorageData.getDataFromLocalStorage('Default Project');
  }
};

// Adding event listener to Project Form
// document.getElementById('projectNameForm').addEventListener('submit', e => {
//   e.preventDefault();
//   submitProjectForm(e.target);
// });
// // Adding event listener to Add Project Btn
// document.getElementById('addProjBtn').addEventListener('click', e => {
//   const action = 'new';
//   showProjectForm(e.target, action);
// });
