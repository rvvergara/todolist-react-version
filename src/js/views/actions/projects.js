import uuid from 'uuid';

const getProjects = projects => ({
  type: 'GET_PROJECTS',
  projects,
});

const addProject = ({ name = 'PROJECT', description = 'SOME PROJECT' } = {}) => ({
  type: 'ADD_PROJECT',
  project: {
    id: uuid(),
    name,
    description,
    todos: [],
  },
});

const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id,
  updates,
});

const deleteProject = id => ({
  type: 'DELETE_PROJECT',
  id,
});

export {
  getProjects,
  addProject,
  updateProject,
  deleteProject,
};
