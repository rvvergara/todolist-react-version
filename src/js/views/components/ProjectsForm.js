import React from 'react';

const ProjectsForm = (props) => {
  const {
    addProjectMode,
    handleChange,
    submitProjectForm,
    editProjectMode,
    name,
    error,
  } = props;
  const displayClass = addProjectMode || editProjectMode ? '' : 'd-none';
  return (
    <div>
      {error && (
      <div>
        { error }
      </div>
      )}
      <form
        className={displayClass}
        onSubmit={submitProjectForm}
      >
        <input
          className="form-control"
          type="text"
          name="projectName"
          value={name}
          onChange={e => handleChange(e.target.name, e.target.value)}
          placeholder={name || 'New Project Name'}
        />
      </form>
    </div>
  );
};

export default ProjectsForm;
