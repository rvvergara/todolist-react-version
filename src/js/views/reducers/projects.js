export default (state = [], action) => {
  switch (action.type) {
    case 'GET_PROJECTS':
      return [...state, ...action.projects]

    case 'ADD_PROJECT':
    {
      const index = state.findIndex(project => project.name === action.project.name);
      return index === -1 ? state.concat(action.project) : state;
    }

    case 'UPDATE_PROJECT':
    {
      const { name } = action.updates;
      const conflictIndex = state.findIndex(project => project.name === name);
      const projectIndex = state.findIndex(project => project.id === action.id);
      const updatedProject = { ...state[projectIndex], ...action.updates };
      return conflictIndex === -1 ? [...state.slice(0, projectIndex), updatedProject, ...state.slice(projectIndex + 1)] : state;
    }
    case 'DELETE_PROJECT':
      return state.filter(project => project.id !== action.id);
    default:
      return state;
  }
};
