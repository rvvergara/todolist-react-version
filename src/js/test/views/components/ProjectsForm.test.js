import React from 'react';
import { shallow } from 'enzyme';
import ProjectsForm from '../../../views/components/ProjectsForm';
import projects from '../../fixtures/projects';

describe('ProjectsForm', () => {
  let handleChange;
  let submitProjectForm;
  let wrapper;

  beforeEach(() => {
    handleChange = jest.fn((k, v) => ({ [k]: v }));
    submitProjectForm = jest.fn();
    wrapper = shallow(
      <ProjectsForm
        handleChange={handleChange}
        submitProjectForm={submitProjectForm}
        name=""
        error=""
      />,
    );
  });

  test('it should render correctly without name and error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('it should render correctly with name', () => {
    wrapper.setProps({ name: 'Another Project' });
    expect(wrapper).toMatchSnapshot();
  });

  test('it should render correctly with error', () => {
    wrapper.setProps({ error: 'You missed something' });
    expect(wrapper).toMatchSnapshot();
  });

  describe('input change and form submission', () => {
    let name;
    let value;
    beforeEach(() => {
      name = 'projectName';
      value = 'Make a change';
      wrapper.find('input').prop('onChange')({
        target: { name, value },
      });
    });
    test('it should handleChange in project name input', () => {
      expect(handleChange).toHaveBeenLastCalledWith(name, value);
    });

    test('it should handle form submission', () => {
      wrapper.find('form').simulate('submit');
      expect(submitProjectForm).toHaveBeenCalled();
    });
  });
});
