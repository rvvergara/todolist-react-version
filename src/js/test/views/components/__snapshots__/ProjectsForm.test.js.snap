// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsForm it should render correctly with error 1`] = `
<div>
  <div>
    You missed something
  </div>
  <form
    onSubmit={[MockFunction]}
  >
    <input
      className="form-control"
      name="projectName"
      onChange={[Function]}
      placeholder="New Project Name"
      type="text"
      value=""
    />
  </form>
</div>
`;

exports[`ProjectsForm it should render correctly with name 1`] = `
<div>
  <form
    onSubmit={[MockFunction]}
  >
    <input
      className="form-control"
      name="projectName"
      onChange={[Function]}
      placeholder="New Project Name"
      type="text"
      value="Another Project"
    />
  </form>
</div>
`;

exports[`ProjectsForm it should render correctly without name and error 1`] = `
<div>
  <form
    onSubmit={[MockFunction]}
  >
    <input
      className="form-control"
      name="projectName"
      onChange={[Function]}
      placeholder="New Project Name"
      type="text"
      value=""
    />
  </form>
</div>
`;
